### PR TITLE
Bug 5363: Handle IP-based X.509 SANs better

### PR DESCRIPTION
**Pull request by walkert**
Most X.509 Subject Alternate Name extensions encountered by Squid are
based on DNS domain names. However, real-world servers (including
publicly available servers that use vanity IP addresses) also use
IP-based SANs. Squid mishandled IP-based SANs in several ways:

* When generating certificates for servers targeted by their IP
  addresses, addAltNameWithSubjectCn() used that target IP as a
  DNS-based SAN, resulting in a frankenstein DNS:[ip] SAN value that
  clients ignored when validating a Squid-generated certificate.

* When validating a received certificate, Squid was ignoring IP-based
  SANs. When Subject CN did not match the requested IP target, Squid
  only looked at DNS-based SANs, incorrectly failing validation.

* When checking certificate-related ACLs like ssl::server_name,
  matchX509CommonNames() ignored IP-based SANs, not matching
  certificates containing ACL-listed IP addresses.

Squid now recognizes and generates IP-based SANs.

Squid now attempts to match IP-based SANs with ACL-listed IP addresses,
but the success of that attempt depends on whether ACL IP parameters are
formatted the same way inet_ntop(3) formats those IP addresses: Matching
is still done using c-string/domain-based ::matchDomainName() (for
ssl::server_name) and string-based regexes (for ssl::server_name_regex).
Similar problems affect dstdomain and dstdomain_regex ACLs. A dedicated
fix is needed to stop treating IPs as domain names in those contexts.

This change introduces partial support for preserving IP-vs-domain
distinction in parsed/internal Squid state rather than converting both
to a string and then assuming that string is a DNS domain name.

### CODE REVIEW

**Comment by rousskov** [comment]
@walkert, thank you for sharing these changes and detailing the logic behind them! I have not reviewed PR code changes yet, but I do have one higher-level question: In your opinion, does this PR address (or at least relates to) [Bug 5363](https://bugs.squid-cache.org/show_bug.cgi?id=5363)? If this PR addresses the same problem, then which solution is better, the patch posted for that bug report or this PR?

**Comment by walkert** [comment]
Thanks @rousskov. So, as far as I can see, the first commit in this PR fixes the bug you're referring to. That bug however does not resolve the second problem we saw which is fixed in the most-recent commit. In terms of which solution is better I would say it may depend on your appetite for changing existing code. My commit modifies the behaviour of the existing `addAltNameWithSubjectCn` to deal appropriately with IPs and DNS names. The patch from that bug introduces a new function `addAltNameWithIpCn`. Further, my first commit (which fixes the bug) introduces a helper function which is relied upon by the most recent commit. Looking again at the patch, the new `addAltNameWithIpCn` contains a fair amount of duplicative code from `addAltNameWithSubjectCn` and I'm not sure whether that's really necessary. Needless to say I'm open to whatever your preference is.

**Comment by rousskov** [comment]
@walkert, thank you for that analysis. We will proceed with your PR as the primary solution, at least for now. I hope to be able to review it, at least superficially next week.

**Comment by walkert** [comment]
Thanks @rousskov. I just noticed a failure due to my forgetting to declare one of the functions as static. Just fixed and pushed that change (then force-pushed a rebase of origin/master).

**Comment by rousskov** [comment]
> (then force-pushed a rebase of origin/master).

FYI: There is no need to rebase (and force push) unless there are merge conflicts or other rare complications. [Anubis](https://github.com/measurement-factory/anubis/blob/master/README.md), the merge bot used by Squid repository, will squash-merge all the changes into master (using PR title/description as the commit message) at the end of the [merge process](https://wiki.squid-cache.org/MergeProcedure).

File: src/ssl/gadgets.cc
Comment (rousskov) [inline src/ssl/gadgets.cc:510]: Please see if you can use Ip::Address::operator "=" instead of calling is_ip_address() here. Ip::Address::operator "=" API is terrible, but it is existing code that we do not want to semi-duplicate unless absolutely necessary (and we would have to semi-duplicate it differently if we have to do that at all).

N.B. This PR adds is_ip_address() that uses inet_pton(). Latest Squid code does not use inet_pton(). It uses Ip::Address::lookupHostIP(...,true) callers. That private Ip::Address method currently uses getaddrinfo().

File: src/ssl/support.cc
Comment (rousskov) [inline src/ssl/support.cc:201]: Please see if you can use Ip::Address::operator "=" instead of calling is_ip_address() here. If the assignment returns true, the resulting Ip::Address object should be usable for comparison.

File: src/ssl/support.cc
Comment (rousskov) [inline src/ssl/support.cc:244]: A few lines above, X509_NAME_get_entry() extracts (what looks like) a CN and then assumes that the extracted name should be compared as a domain name. Is it possible that the extracted name is an IP address? PR addAltNameWithSubjectCn() code looks similar and assumes the extracted NID_commonName can be an IP. If the answer is "yes, it is possible", then this PR should adjust the above code to handle IPs, right? Otherwise, please add a source code comment explaining why IP addresses are not possible in that contexts (or possible but should be ignored).

I cannot attach this change request to that currently-unmodified code.

File: src/ssl/support.cc
Comment (rousskov) [inline src/ssl/support.cc:260]: This GEN_IPADD `case` addition feels like a Ssl::matchX509CommonNames() API violation to me. The function promises to (and did) outsource "yes, there is a match" conclusion to check_func. This PR-added code ignores that promise.

Ssl::matchX509CommonNames() API is used by several code paths. I suspect that many (if not all) of those paths would benefit from being able to handle GEN_IPADD entries.

I suspect that we have to adjust Ssl::matchX509CommonNames() API to replace check_func function with a C++ class object that has two methods: matchDomainName() and matchIp().

File: src/ssl/support.cc
Comment (rousskov) [inline src/ssl/support.cc:219]: I hope that after Ip::Address::operator "=" is used to parse check_ip, we would be able to use existing Ip::Address comparison methods instead of raw memcmp(). Again, that existing code has many problems, but we do not want to semi-duplicate it unless we really have to.

**Review by rousskov — CHANGES_REQUESTED** [review]
I have left a few high-level change requests. Once those are addressed, I can help polish this code to meet various Squid code style [guidelines](https://wiki.squid-cache.org/DeveloperResources/SquidCodingGuidelines).

File: src/ssl/gadgets.cc
Comment (walkert) [reply src/ssl/gadgets.cc:510]: Hi @rousskov. I wonder if you could help please. I'm attempting to use `Ip::Address::operator` by adding `#include ip/Address.h` to gadgets.cc. I then do something like this:
```
Ip::Address ipAddress;
if (ipAddress = address_str) {
  type = "IP"
} else {
  type = "DNS"
}
```
After running `make` I get this error in the output:

```
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `Ip::Address::Address()':
/tmp/squid-6.9/src/ssl/../../src/ip/Address.h:46: undefined reference to `Ip::Address::setEmpty()'
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `addAltNameWithSubjectCn(Security::LockingPointer<x509_st, &Security::X509_free_cpp, HardFun<int, x509_st*, &X509_up_ref> >&)':
/tmp/squid-6.9/src/ssl/gadgets.cc:490: undefined reference to `Ip::Address::operator=(char const*)'
collect2: error: ld returned 1 exit status
```

I've tried to use a couple of the `Ip::Address` methods which resulted in the same error. Would you be able to point me in the right direction to fix this please?

Thanks

Comment (rousskov) [reply src/ssl/gadgets.cc:510]: Shared output does not detail which executable fails to link, but I am guessing it is security_file_certgen. See the difference between two comments describing two libraries in src/ssl/Makefile.am. Compared to the primary squid executable, security_file_certgen helper is build using a (smaller but naturally growing) subset of libraries; we split src/ssl code into two libraries to minimize the number of (indirect) security_file_certgen dependencies[^1].

If my guess is correct, then try adding `$(top_builddir)/src/ip/libip.la` to `security_file_certgen_LDADD` in src/security/cert_generators/file/Makefile.am. It should probably go on the first or second line, above or below libsslutil.la. You may they need to add more libraries if libip.la has unresolved dependencies. Do not forget to re-bootstrap after modifying an Automake file!

If the suggestion above works, then the original two-library split described above will no longer be necessary (but removing it should be done in a dedicated PR so do not worry about it for now).

[^1]: We plan to migrate away from that painful minimization approach, but the code you have to deal with has not been upgraded, so you have to link security_file_certgen (and other binaries) with the necessary convenience libraries (often discovered by looking at linker errors like the one you have shared).

Comment (walkert) [reply src/ssl/gadgets.cc:510]: You were right. However, I've made the change you suggested to `src/security/cert_generators/file/Makefile.am` then I ran `./bootstrap.sh` followed by `./configure` from `$(top_builddir)` but I see no change in the error reported. Here it is with a bit more context:

```
make[4]: Entering directory '/portal/squid-6.9/src/security/cert_generators'                                                      
Making all in file                                                                                                                
make[5]: Entering directory '/portal/squid-6.9/src/security/cert_generators/file'
/bin/bash ../../../../libtool  --tag=CXX   --mode=link g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -I/usr/include/p11-kit-1    -g -O2 -march=native  -g -o security_file_certgen certificate_db.o security_file_certgen.o ../../../../src/ip/libip.la ../../../../src/ssl/libsslutil.la ../../../../src/sbuf/libsbuf.la ../../../../src/debug/libdebug.la ../../../../src/error/liberror.la ../../../../src/comm/libminimal.la ../../../../src/mem/libminimal.la ../../../../src/base/libbase.la ../../../../src/time/libtime.la -lssl -lcrypto   -lgnutls   ../../../../compat/libcompatsquid.la                                                                      
libtool: link: g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -I/usr/include/p11-kit-1 -g -O2 -march=native -g -o security_file_certgen certificate_db.o security_file_certgen.o  ../../../../src/ip/.libs/libip.a ../../../../src/ssl/.libs/libsslutil.a ../../../../src/sbuf/.libs/libsbuf.a ../../../../src/debug/.libs/libdebug.a ../../../../src/error/.libs/liberror.a ../../../../src/comm/.libs/libminimal.a ../../../../src/mem/.libs/libminimal.a ../../../../src/base/.libs/libbase.a ../../../../src/time/.libs/libtime.a -lssl -lcrypto -lgnutls ../../../../compat/.libs/libcompatsquid.a                                                                   
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `Ip::Address::Address()':
/portal/squid-6.9/src/ssl/../../src/ip/Address.h:46: undefined reference to `Ip::Address::setEmpty()'
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `addAltNameWithSubjectCn(Security::LockingPointer<x509_st, &Security::X509_free_cpp, HardFun<int, x509_st*, &X509_up_ref> >&)':
/portal/squid-6.9/src/ssl/gadgets.cc:490: undefined reference to `Ip::Address::operator=(char const*)'
collect2: error: ld returned 1 exit status        
```

It seems like libsslutil is not seeing the link but I'm afraid I'm not sure how to proceed from here either if you can suggest please. Appreciate your help/patience.

Comment (rousskov) [reply src/ssl/gadgets.cc:510]:
```
... security_file_certgen.o ../../../../src/ip/libip.la ../../../../src/ssl/libsslutil.la ...
```

Try reordering the first two libraries. In other words, place libip _after_ libsslutil in Makefile.am.

Rule of thumb: Libraries that provide symbol X should be listed _after_ libraries that use symbol X.

Comment (walkert) [reply src/ssl/gadgets.cc:510]: Ah thanks. I had tried it below with a different source checkout and was sure I'd seen errors but doing as you suggest I was able to get `make` to complete. I'll now see about making some of these changes. Thanks.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc:244]: Yes, it's possible. The issue in my case is that even though the extracted name _is_ an IP address, it isn't a match and thus we must go through the SANs until we get a valid comparison. Your point is of course valid though, if we make `*check_func` handle both types of comparison, the code below can be simplified to just remove the skipping of non-GEN_DNS entries. I'll look into a modification of the check_func code.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc:305]: So, I've added 2 new commits. The most recent adds a `VerifyAddress` class. However, so that I don't have to mess with the calling semantics for matchX509CommonNames in other locations, I've retained the API to expect a `check_func` (or equivalent) but modified the body of the original `check_func` to use the new `VerifyAddress::match` method.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: Yes, I've completely removed this function now and used functionality from Ip::Address. Thanks!

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: Updated commits now use the `=` for comparison.

File: src/security/cert_generators/file/Makefile.am
Comment (walkert) [inline src/security/cert_generators/file/Makefile.am:27]: Required to allow gadgets.cc to refer to `ip/Address.h`.

File: src/ssl/support.cc
Comment (walkert) [inline src/ssl/support.cc:100]: Note that all logic from this method is copied from the original `check_domain` function.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc:305]: FWIW - I found it more simple to retain the `switch` and then ensure the correct `ASN1_STRING` is passed to `match`. So we're still outsourcing the check/conclusion to check_func but being purposeful about which data we send to it for comparison.

File: src/ssl/support.cc
Comment (rousskov) [reply src/ssl/support.cc:289]: I assume that the check function is now expected to handle both IP addresses and domain names. Thus, recent PR modifications address this change request. This PR will need to _document_ that expectation (probably inside that check function type description), and there may be concerns about check function ability to correctly determine the type of information it is given, but I am resolving _this_ change request as addressed. Thank you.

File: src/ssl/support.cc
Comment (rousskov) [inline src/ssl/support.cc:303]: N.B. The type of the second check_func() argument is (a pointer to) an ASN1_STRING. "Almost all ASN1 types in OpenSSL are represented as an ASN1_STRING structure. ASN1_STRING is also used for some CHOICE types." Roughly speaking, ASN1_STRING can store nearly anything.

I am concerned that this code loses "this ASN1_STRING is a GEN_IPADD SAN (rather than, say, GEN_DNS SAN)" knowledge and, hence, forces check_func() to _guess_ what information its second parameter actually contains. That guess is currently based on ASN1_STRING_type(asn1_data) not being V_ASN1_OCTET_STRING. I assume that other CN and SAN types may pass that rather weak "type check", in current or future code, resulting in silently incorrect casting and incorrect matching in check_func() calls that are not as well protected as this one.

I do understand that you would prefer not to fiddle with other check_func callers. However, this information loss and associated dangerous guesses is a very large red flag for me. If I am missing some built-in protections that make this loss and guesses safe, please do let me know. If this concern is not addressed (one way or the other), I will find the time to start addressing it myself after the next iteration, after you address other change requests (that do not require changing other check_func() callers).

**Review by rousskov — DISMISSED** [review]
You have made significant progress, thank you! I left a few change requests for the next iteration. I continue to ignore most of the secondary issues that we will handle later.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc:87]: Ah, I lifted this from the original code in check_domain found [here](https://github.com/squid-cache/squid/blob/master/src/ssl/support.cc#L234). As far as I can see, the only additional data I'm passing here that the original code is not is when the type is GEN_IPADD, otherwise, the inputs to this should be the same as they were prior to the PR?

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: This code no longer exists.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: The latest commit builds either in_addr or in6_addr as appropriate and then creates an `Ip::Address` from it for comparison. I believe this resolves the issue here though I'm certainly open to a better way of dealing with the IPV4/6 logic.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: Removed.

File: src/ssl/support.cc
Comment (walkert) [reply src/ssl/support.cc]: So in the latest commit, I have indeed fiddled with all of the callers. In all call sites, an explicit enum `Ssl::AddressType` is used to tell the `check_domain` function which address type it is using to remove the guess work. I hope this address the red flag concern.

**Comment by walkert** [comment]
Hi @rousskov, please excuse the significant delay in getting back to you on this. I've finally found the time to continue working on the PR. I've added comments which hopefully explain some of the latest commit. Thanks.

**Comment by rousskov** [comment]
> Hi @rousskov, please excuse the significant delay in getting back to you on this. I've finally found the time to continue working on the PR. I've added comments which hopefully explain some of the latest commit.

Thank you very much for making progress with this PR. Now it is my turn to find the time to get back to this project. It will _not_ happen in the next few days. Please do not hesitate to ping me if I do not respond by July 18, 2024.

**Comment by walkert** [comment]
Hi @rousskov hope all is well. Per your message, I'm pinging you as we've passed the 18th. Thanks!

**Comment by rousskov** [comment]
> Per your message, I'm pinging you as we've passed the 18th. Thanks!

Thank you! Working on it now.

**Review by rousskov — COMMENTED** [review]
_Status update_: I have completed a high-level review of current PR code. I believe the guts of the necessary changes are there (thank you!). There are some dangerous conversions, tricky code duplication, misplaced code, and C-style complications that I am working on addressing now.

@walkert, I expect to push a few major changes in the next few days. If they pass your tests and review, then I will help you polish PR code to satisfy Squid code style guidelines and merge this PR. It should not take long. For now, the ball is still on my side.

**Comment by rousskov** [comment]
> I expect to push a few major changes in the next few days. 

Done. I tried to document the rationale behind most changes in the corresponding commit messages. There is one major problem that I have not solved (see XXX in Ssl::OneNameMatcher::matchIp()), but I doubt its solution, whatever it will be, is going to require serious refactoring of other PR changes. Some code and comments need polishing, but that can wait for the next iteration.

> If they pass your tests and review, then I will help you polish PR code to satisfy Squid code style guidelines and merge this PR. It should not take long. For now, the ball is still on my side.

@walkert, please _test_ and review. Please feel free to revert all or some of my changes and adjust the current PR code as needed. I am happy to answer any questions. The ball is now on your side.

P.S. Please ignore CI `build-tests` failures for now, but do adjust CONTRIBUTORS to fix `source-maintenance-tests` [failure](https://github.com/squid-cache/squid/actions/runs/10169661501/job/28127084151?pr=1793) (by picking the email variation you prefer).

**Comment by walkert** [comment]
This is amazing work @rousskov, thanks so much for your effort. I have tested and can confirm this still works as expected. I have also updated CONTRIBUTORS as requested. What should I do next?

**Comment by rousskov** [comment]
> I have tested and can confirm this still works as expected. I have also updated CONTRIBUTORS as requested. What should I do next?

Thank you for a quick turnaround! Give me a couple of days to see if I can solve XXX in Ssl::OneNameMatcher::matchIp() -- I have an idea that I want to test. After that (and some minor code polishing), this PR should be ready for wider review. The ball is on my side again.

**Comment by rousskov** [comment]
> Give me a couple of days to see if I can solve XXX in Ssl::OneNameMatcher::matchIp()

Done. This PR is now free of known in-scope XXXs and TODOs.

I also updated PR title/description, primarily to better reflect the (longer) list of fixed bugs and to disclose unresolved (out of scope) issues. @walkert, when you get a chance, please review the current PR title/description and let me know if I missed or misrepresented anything important.

**Comment by walkert** [comment]
Thanks again for all your work @rousskov. I'm not sure this can even be considered my PR anymore! Regardless, I've reviewed your changes to the title and description and it all looks good to me.

File: src/anyp/Host.h
Comment (rousskov) [inline src/anyp/Host.h:24]: This class is at the core of this PR changes. In official code, a single address that may be either an IP address or a DNS domain name is either isolated into two or more co-existing objects (e.g., host information inside AnyP::Uri) or morphed into one opaque SBuf object (e.g., `server` in ssl_verify_cb()). Neither approach works well, for obvious reasons. This PR uses the new Host class to accurately and as-safely-as-possible represent this naming duality in several X.509 SAN-related contexts. Eventually, more code should be converted to use this class. I have left a few related TODOs.

The proposed "host" name works well enough IMO, but I would not be surprised if there are better alternatives our there.

File: src/dns/forward.h
Comment (rousskov) [inline src/dns/forward.h:41]: For now, this alias allows us to document Dns::DomainName properties/invariants without changing a lot of code to cope with a new dedicated "real" type. It might stay that way, but I would not be surprised if we will eventually convert this into a class to reduce chances of wrong information (e.g., an IP address or even address:port authority) sneaking into this object and screwing up ACL checks and such.

Technically, this type alias declaration is not a true C++ forward declaration. I think we should make an exception for this forward.h entry, at least for now, because there is no better place to put it. However, if others insist, we can add something like src/dns/elements.h and put it there.

File: src/ssl/support.h
Comment (rousskov) [inline src/ssl/support.h:318]: Since we are not really finding the position of the given needle here, we may want to use a hasFoo() naming scheme instead:

```suggestion
bool hasSubjectName(X509 &, const AnyP::Host &);
```

Same for findMatchingSubjectName(), of course.

We should also capitalize these free functions to follow Squid naming guidelines. I will do that once the name is settled.

File: src/ssl/support.h
Comment (rousskov) [inline src/ssl/support.h:285]: This C-style function was one of the primary sources of problems fixed in this PR. As you can see, it was using essentially two void* pointers (ASN1_STRING can be nearly anything) and expected the passed callback (i.e. check_func) to figure out what X.509 name it is being called with. The replacement GeneralNameMatcher API (declared below) provides strong visual clues for developers backed by compile-time type checks to make sure we are not trying to compare a domain name with an IP address or simply forget to handle an IP address case.

**Review by rousskov — COMMENTED** [review]
I think this PR is ready for review by others (CC: @yadij, @kinkie). The original triage, initial fix implementation, and testing belong to @walkert. All bugs are mine. I left a few comments to assist with reviews. If you want any of this PR changes factored out into prerequisite PRs (without increasing their breadth/scope), please let me know, and I will help with that.

Once the dust settles, we can merge from master to fix CI and rename a few functions as detailed in one of my comments. These pending changes should not affect any reviews.

If there are no objections, I plan to clear this PR for merging in ten calendar days.

**Comment by rousskov** [comment]
> Alex: If there are no objections, I plan to clear this PR for merging in ten calendar days.

Status update: Amos has self-requested a review, so the ball is on his side, and the timeout I penciled in above is now irrelevant.

**Comment by walkert** [comment]
Hi @yadij. I just wanted to check-in and see if there's anything I can do to help with this review? Thanks.

File: src/anyp/Host.h
Comment (walkert) [reply src/anyp/Host.h:24]: Hi @rousskov - do you know if there's anything else that needs to be done for the review to continue? Thanks.

Comment (rousskov) [reply src/anyp/Host.h:24]: @walkert, unfortunately, the answer to your question is "no": I do not know how to convince @yadij to review PRs awaiting his review. IMO, this persistent problem is one of the existential threats faced by the Squid Project, but all my attempts to find a solution have failed. I can only hope that we will resume making progress with this PR soon.

Comment (walkert) [reply src/anyp/Host.h:24]: @rousskov understood. Thanks for taking the time to respond. I'll keep watching and hoping.

Comment (walkert) [reply src/anyp/Host.h:24]: Hi @rousskov, is there scope to hand this off to another reviewer if @yadij is unavailable? Thanks.

Comment (rousskov) [reply src/anyp/Host.h:24]: @walkert, there is no formal procedure to overwrite, move, or clear a "this PR needs to be reviewed by me" flag set by a core developer. Project core developers hold a lot of power, and multiple attempts to establish timeouts (or otherwise restructure the governing rules) have failed so far. FWIW, I plan to bring this PR up again during the next meeting which should happen on October 22, 2024.

Comment (walkert) [reply src/anyp/Host.h:24]: @rousskov understood. Thanks for the info. 🙏

File: src/ssl/support.h
Comment (rousskov) [reply src/ssl/support.h]: Done in commit a8e992a0.

**Review by rousskov — APPROVED** [review]
I accidentally noticed that Amos has removed his request for review and cleared this PR for merging. I now finished last in-scope TODO, merged from master to get the current set of CI tests (that have changed quite a bit while this PR was waiting), and approved this PR so that it can be auto-merged by Anubis (if those tests succeed).

@walkert, please ping me if you notice any failed tests.

**Comment by walkert** [comment]
Thanks again for all your work @rousskov. I'll keep my eyes peeled.

**Comment by rousskov** [comment]
@walkert, you may be interested in #1967 that attempts to fix a bug I introduced in this PR. If you can test that fix, please do.

**Comment by walkert** [comment]
> @walkert, you may be interested in #1967 that attempts to fix a bug I introduced in this PR. If you can test that fix, please do.

@rousskov apologies for the lengthy delay in replying. I haven't had a chance to look at this but will try and do so soon.

**Pull request closed without merging**
